experimental/lib/Transforms/LSQSizing/HandshakeSizeLSQs.cpp

```diff
@@ -0,0 +1,672 @@
+
+  // Finds the allocation time of each operation, which is the earliest start
+  // time of the Phi Node of the BB plus a fixed additional latency
+  TimePerOpMap
+  getAllocTimes(AdjListGraph graph, mlir::Operation *startNode,
+                std::vector<mlir::Operation *> ops,
+                std::unordered_map<unsigned, mlir::Operation *> phiNodes);
+
+  // Finds the deallocation time of each store operation, which is the the
+  // latency of the last argument arriving plus a fixed additional latency
+  TimePerOpMap getStoreDeallocTimes(AdjListGraph graph,
+                                    mlir::Operation *startNode,
+                                    std::vector<mlir::Operation *> storeOps);
+
+  // Finds the deallocation time of each load operation, which is the latest
+  // argument arriving at the operation succeding the load operation plus a
+  // fixed additional latency This is due to to the fact that the load operation
+  // frees the queue entry as soon as the load result is passed on to the
+  // succeeding operation
+  TimePerOpMap getLoadDeallocTimes(AdjListGraph graph,
+                                   mlir::Operation *startNode,
+                                   std::vector<mlir::Operation *> loadOps);
+
+  // Given the alloc and dealloc times of each operation, calculates the maximum
+  // queue size needed for each LSQ
+  SizePerOpMap
+  calcQueueSize(std::unordered_map<unsigned, TimePerOpMap> allocTimes,
+                std::unordered_map<unsigned, TimePerOpMap> deallocTimes,
+                std::vector<unsigned> IIs);
+};
+} // namespace
+
+void HandshakeSizeLSQsPass::runDynamaticPass() {
+  llvm::SmallVector<LSQSizingResult> sizingResults;
+
+  // Read component latencies
+  TimingDatabase timingDB(&getContext());
+  if (failed(TimingDatabase::readFromJSON(timingModels, timingDB)))
+    signalPassFailure();
+
+  mlir::ModuleOp mod = getOperation();
+  for (handshake::FuncOp funcOp : mod.getOps<handshake::FuncOp>()) {
+
+    std::unordered_map<unsigned, llvm::SetVector<unsigned>> cfdfcBBLists;
+    std::unordered_map<unsigned, float> IIs;
+
+    // Extract CFDFCs and II from the Attributes
+    handshake::CFDFCThroughputAttr throughputAttr =
+        getDialectAttr<handshake::CFDFCThroughputAttr>(funcOp);
+    handshake::CFDFCToBBListAttr cfdfcAttr =
+        getDialectAttr<handshake::CFDFCToBBListAttr>(funcOp);
+
+    if (throughputAttr == nullptr || cfdfcAttr == nullptr) {
+      continue;
+    }
+
+    DictionaryAttr throughputDict = throughputAttr.getThroughputMap();
+    DictionaryAttr cfdfcDict = cfdfcAttr.getCfdfcMap();
+
+    // Convert Attribute into better usable data structure
+    for (auto &attr : cfdfcDict) {
+      ArrayAttr bbList = llvm::dyn_cast<ArrayAttr>(attr.getValue());
+      llvm::SetVector<unsigned> cfdfcBBs;
+      for (auto &bb : bbList) {
+        cfdfcBBs.insert(bb.cast<IntegerAttr>().getUInt());
+      }
+      cfdfcBBLists.insert({std::stoi(attr.getName().str()), cfdfcBBs});
```

Comment:
Use `StringRef::getAsInteger` method instead of `std::stoi` to gracefully catch eventual parsing errors (`std::stoi` will just throw an exception in case of error, crashing the program). Here that would probably look like

```cpp
unsigned index;
if (attr.getName().getAsInteger(10, index)) {
  // report an error
}
// everything is fine, you can use index
```

Reply:
What would be the best way to report the error?
1. printing something with `llvm::errs()`, skipping the CFDFC and then continue 
2. calling `signalPassFailure()`
3. something else?

Reply:
Since you are in `runDynamaticPass` doing `return signalPassFailure();` seems like the best option here.